common/src/main/java/com/wynntils/features/user/ChatMentionFeature.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.user;
+
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.mc.McUtils;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatMentionFeature extends UserFeature {
+    @Config
+    public boolean markMention = true;
+
+    @Config
+    public boolean dingMention = true;
+
+    @Config
+    ChatFormatting rewriteColorCode = ChatFormatting.YELLOW;
+
+    @Config
+    public String aliases = "";
+
+    private Pattern pattern;
+
+    @Override
+    protected void onConfigUpdate(ConfigHolder configHolder) {
+        // rebuild pattern in case it has changed
+        buildPattern();
+    }
+
+    private void buildPattern() {
+        pattern = Pattern.compile(
+                "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                        + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+                Pattern.CASE_INSENSITIVE);
+    }
+
+    @SubscribeEvent()
+    public void onChat(ChatMessageReceivedEvent e) {
+        Component message = e.getMessage();
+
+        Matcher looseMatcher = pattern.matcher(ComponentUtils.getUnformatted(message));
+
+        if (looseMatcher.find()) {
+            if (markMention) {
+                e.setMessage(rewriteComponentTree(message));
+            }
+            if (dingMention) {
+                McUtils.playSound(SoundEvents.NOTE_BLOCK_PLING.value());
+            }
+        }
+    }
+
+    private Component rewriteComponentTree(Component comp) {
+        // Make a copy of the component without the siblings
+        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
+        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
+        // because wynn still uses legacy coloring for it.
+        String text = curr.getString();
+
+        // if current component has text check it for mentions
+        if (text != "") {
+            Matcher match = pattern.matcher(text);
+
+            // if we match then rewrite the component
+            if (match.find()) {
+                // get the start of the string before any mention and set curr to be it
+                String before = text.substring(0, match.start());
+                curr = Component.literal(before).withStyle(comp.getStyle());
+
+                int nextStart = 0;
```

Comment:
But, also, after you have moved this code into a separate method, you also need to simplify it further. The looping logic with a mutable nextStart int, a do while loop (which are much harder to read than while loops), and the loop starting with a check for nextStart which can never be != 0 when you start, etc. It's just a mess.

Think like this: It is not enough for the code to **have no obvious faults**. The code should be written so it **obviously have no faults**. Unfortunately, this code falls in the former category, rather than the latter. It looks like it might work, but it is not immediately clear that it does.

Reply:
Alright I've done my best to clean up things as much as possible and at this point I'm all out of ideas of how to clean things up further. If someone else wants to take a crack at cleaning things up more they can go ahead but I'm done with this code at least for now.